File: frontend/src/types/plan.ts
export interface Plan {
  id: string;
  name: string;
  description: string | null;
  sensor_id: string;
  channel: string;
  target_value: number;
  target_unit: string | null;
  lower_limit: number | null;
  upper_limit: number | null;
  enabled: boolean;
}

export interface PlanAction {
  id: string;
  plan_id: string;
  action_type: "notification" | "sms" | "call" | "reminder";
  trigger_on: "above_upper" | "below_lower" | "outside_range" | "any_breach";
  message_template: string;
  recipient_user_ids: string[] | null;
  config: Record<string, string>;
  enabled: boolean;
}

export interface Task {
  id: string;
  title: string;
  description: string | null;
  deadline: string | null;
  status: "open" | "submitted" | "done" | "rejected";
  created_by_id: string | null;
  reviewed_by_id: string | null;
  reviewed_at: string | null;
  review_note: string | null;
  submitted_by_id: string | null;
  submitted_at: string | null;
  completion_note: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskAttachment {
  id: string;
  task_id: string;
  filename: string;
  original_name: string;
  file_type: "image" | "video" | "document" | "other";
  file_size: number | null;
}

export interface TaskDetail extends Task {
  assignees: { id: string; full_name: string; email: string }[];
  attachments: TaskAttachment[];
}

File: backend/app/models/plan.py
import uuid
from datetime import datetime
from sqlalchemy import String, Text, Float, Boolean, DateTime, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column
from app.core.database import Base


class Plan(Base):
    """Metric target plan for a sensor channel."""
    __tablename__ = "plans"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    sensor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("sensors.id", ondelete="CASCADE"), nullable=False
    )
    channel: Mapped[str] = mapped_column(String(100), nullable=False)
    target_value: Mapped[float] = mapped_column(Float, nullable=False)
    target_unit: Mapped[str | None] = mapped_column(String(50), nullable=True)
    lower_limit: Mapped[float | None] = mapped_column(Float, nullable=True)
    upper_limit: Mapped[float | None] = mapped_column(Float, nullable=True)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    created_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class PlanAction(Base):
    """Action to execute when a plan's sensor value breaches limits."""
    __tablename__ = "plan_actions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    plan_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("plans.id", ondelete="CASCADE"), nullable=False
    )
    # "notification" | "sms" | "call" | "reminder"
    action_type: Mapped[str] = mapped_column(String(50), nullable=False)
    # "above_upper" | "below_lower" | "outside_range" | "any_breach"
    trigger_on: Mapped[str] = mapped_column(String(50), nullable=False, default="any_breach")
    message_template: Mapped[str] = mapped_column(
        String(500), nullable=False,
        default="Sensor {sensor} channel {channel}: value {value} breached plan target {target}"
    )
    # NULL = send to all users with sensor access; otherwise specific user IDs
    recipient_user_ids: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # action-specific config: {"phone": "+1555..."} for sms/call
    config: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
